src/mode/modeVisual.ts

```diff
@@ -0,0 +1,140 @@
+"use strict";
+
+import * as _      from 'lodash';
+
+import { ModeName, Mode } from './mode';
+import { Motion} from './../motion/motion';
+import { Position } from './../motion/position';
+import { Operator } from './../operator/operator';
+import { DeleteOperator } from './../operator/delete';
+import { ModeHandler } from './modeHandler.ts';
+import { ChangeOperator } from './../operator/change';
+
+export class VisualMode extends Mode {
+    /**
+     * The part of the selection that stays in the same place when motions are applied.
+     */
+    private _selectionStart: Position;
+
+    /**
+     * The part of the selection that moves.
+     */
+    private _selectionStop : Position;
+    private _modeHandler   : ModeHandler;
+
+    private _keysToOperators: { [key: string]: Operator };
+
+    constructor(motion: Motion, modeHandler: ModeHandler) {
+        super(ModeName.Visual, motion);
+
+        this._modeHandler = modeHandler;
+        this._keysToOperators = {
+            // TODO: use DeleteOperator.key()
+
+            // TODO: Don't pass in mode handler to DeleteOperators,
+            // simply allow the operators to say what mode they transition into.
+            'd': new DeleteOperator(modeHandler),
+            'x': new DeleteOperator(modeHandler),
+            'c': new ChangeOperator(modeHandler)
+        };
+    }
+
+    shouldBeActivated(key: string, currentMode: ModeName): boolean {
+        return key === "v";
+    }
+
+    async handleActivation(key: string): Promise<void> {
+        this._selectionStart = this.motion.position;
```

Comment:
Hmm... Well, to start, I still don't fully grasp the concept of `Motion`. It seems to me like it's just a wrapper around Position to update VSCode, but I think those are actually 2 separate responsibilities that don't need to be in the same class. 

This seems like a good place to discuss some of my thoughts about structuring this code. My idea is to pass around a state object which could look like this (very rough)

```{
    cursorStart: Position(0, 5),
    cursorEnd: Position(0, 7),
    selecting: true,
    mode: Mode.Visual,
    textInEditor: "hello world" // to facilitate undo - maybe stored as diffs to save memory, idk?
}```

All motions would just take a state and return a new state object. When you pressed a key the object would be passed through some transformations and eventually you'd get a new one. When the process is done you would then update VSCode to show the new selections and cursor positions.

The idea is to make stuff like macros/number commands easier - if, say, the user goes into visual mode, makes 2w into a macro and does 100@@, we don't redraw the selection 200 times. 

This would also make proper undo (e.g., stepping back through full actions rather than what we have now) way more straightforward, since actions are nicely discretized. 

Anywho, I still have to dig more into the source of Vintageous and VSVim before I feel confident about making sweeping changes like this, but those are my general thoughts. What do you think?

(I also am a fan of separating out the motions into classes like VSVim and Vintageous do - otherwise the Motion and Position classes would eventually become monstrous!)